### PR TITLE
rocksdb: 6.4.6 -> 6.10.2

### PR DESCRIPTION
**Pull request by prusnak**
###### Motivation for this change

Version update

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by prusnak** [comment]
@GrahamcOfBorg build rocksdb

File: pkgs/development/libraries/rocksdb/default.nix [hunk 1]
```diff
@@ -6,37 +6,26 @@
...
-    (fetchpatch {
-      url = "https://github.com/facebook/rocksdb/commit/c0be4b2ff1a5393419673fab961cb9b09ba38752.diff";
-      sha256 = "1f2wg9kqlmf2hiiihmbp8m5fr2wnn7896g6i9yg9hdgi40pw30w6";
-    })
-  ];
-
-  postPatch = ''
-    substituteInPlace CMakeLists.txt --replace "find_package(zlib " "find_package(ZLIB "
-  '';
+  propagatedBuildInputs = [ bzip2 lz4 snappy zlib zstd ];
```
Comment (danieldk) [inline pkgs/development/libraries/rocksdb/default.nix:19]: What's the reason to propagate these dependencies? They don't seem to leak through public rocksdb headers?

Comment (prusnak) [reply pkgs/development/libraries/rocksdb/default.nix:19]: See https://github.com/NixOS/nixpkgs/pull/82690#issuecomment-600400102 and the subsequent comments.

Also looking at packages that use rocksdb as a dependency, they seem to add these into their `buildInputs` indicating (at least some of) these really need to be included in the build:

* https://github.com/NixOS/nixpkgs/blob/9b195c5369b906825134aafce59744e89b85bd37/pkgs/servers/blockbook/default.nix#L29
* https://github.com/NixOS/nixpkgs/blob/9b195c5369b906825134aafce59744e89b85bd37/pkgs/tools/filesystems/ceph/default.nix#L148

Comment (danieldk) [reply pkgs/development/libraries/rocksdb/default.nix:19]: I am wondering if these are actual dependencies caused by rocksdb. rocksdb does not export zstd types. So either downstream packages are linking against the static library or something else strange is going on.

* Uses lz4 and snappy itself, so these dependencies probably do not come from its rocksdb dependency (unless the version of rocksdb that is vendored as a submodule is compiled).
* I didn't check the full dependency graph for blockbook, but the Go module that is used as a rocksdb bindings does not use e.g. snappy. I think there may just be other dependencies that use these libraries.
* sortmerna currently build fine without these dependencies?

Comment (danieldk) [reply pkgs/development/libraries/rocksdb/default.nix:19]: Oh, I guess that `sortmerna` only failed after the version bump? Let me check that!

Comment (danieldk) [reply pkgs/development/libraries/rocksdb/default.nix:19]: > Oh, I guess that `sortmerna` only failed after the version bump? Let me check that!

That's a problem with `sortmerna`'s CMake file, it works fine with the latest version of `sortmerna`. I'll do a PR to update `sortmerna`.

Comment (prusnak) [reply pkgs/development/libraries/rocksdb/default.nix]: OK, I updated this PR to not change `buildInputs` to `propagatedBuildInputs` in b443d7ecf44cd753b8ceaab4fa891cfd09325c5c

Comment (danieldk) [reply pkgs/development/libraries/rocksdb/default.nix]: Scratch that, it was not `sortmerna`'s cmake files. But I have found the real culprit. `sortmerna` uses `rocksdb`'s CMake file, which also contains:

```
set_target_properties(RocksDB::rocksdb-shared PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "/nix/store/776p1v120ij0vhv5xdb65g0qqd6wxbf2-rocksdb-6.10.2/include"
  INTERFACE_LINK_LIBRARIES "snappy::snappy;ZLIB::ZLIB;/nix/store/nsfb8xdr47sxsfsa2bp0zf1kcm959z48-bzip2-1.0.6.0.1/lib/libbz2.so;lz4::lz4;zstd::zstd;-lpthread"
)
```

This is wrong, `INTERFACE_LINK_LIBRARIES` should not be there. I have a fix for this. Do you mind if I push it to this PR?

**Comment by prusnak** [comment]
@GrahamcOfBorg build rocksdb

File: pkgs/development/libraries/rocksdb/default.nix (hunk 1 above)
Comment (danieldk) [reply pkgs/development/libraries/rocksdb/default.nix]: This is not yet resolved. The build creates incorrect CMake files. You need to apply this diff:

```diff
diff --git a/pkgs/development/libraries/rocksdb/cmake-private-third-party.diff b/pkgs/development/libraries/rocksdb/cmake-private-third-party.diff
new file mode 100644
index 00000000000..9572a40026e
--- /dev/null
+++ b/pkgs/development/libraries/rocksdb/cmake-private-third-party.diff
@@ -0,0 +1,19 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index 302575cb7..6276972ef 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -839,12 +839,12 @@ else()
+ endif()
+ 
+ add_library(${ROCKSDB_STATIC_LIB} STATIC ${SOURCES})
+-target_link_libraries(${ROCKSDB_STATIC_LIB}
++target_link_libraries(${ROCKSDB_STATIC_LIB} PRIVATE
+   ${THIRDPARTY_LIBS} ${SYSTEM_LIBS})
+ 
+ if(ROCKSDB_BUILD_SHARED)
+   add_library(${ROCKSDB_SHARED_LIB} SHARED ${SOURCES})
+-  target_link_libraries(${ROCKSDB_SHARED_LIB}
++  target_link_libraries(${ROCKSDB_SHARED_LIB} PRIVATE
+     ${THIRDPARTY_LIBS} ${SYSTEM_LIBS})
+ 
+   if(WIN32)
diff --git a/pkgs/development/libraries/rocksdb/default.nix b/pkgs/development/libraries/rocksdb/default.nix
index fcf2190672d..17dad514d4c 100644
--- a/pkgs/development/libraries/rocksdb/default.nix
+++ b/pkgs/development/libraries/rocksdb/default.nix
@@ -15,6 +15,10 @@ stdenv.mkDerivation rec {
     sha256 = "1f2wqb6px812ijcivq3rsknqgkv01wblc6sd8wavhrw8qljgr3s1";
   };
 
+  patches = [
+    ./cmake-private-third-party.diff
+  ];
+
   nativeBuildInputs = [ cmake ninja ];
   propagatedBuildInputs = [ bzip2 lz4 snappy zlib zstd ];
 
```

Comment (prusnak) [reply pkgs/development/libraries/rocksdb/default.nix]: I assume your changes are extracted from this change, right? https://github.com/facebook/rocksdb/commit/07204837ce8d66e1e6e4893178f3fd040f9c1044

In that case, I'll fetch the patch from GitHub instead.

Comment (prusnak) [reply pkgs/development/libraries/rocksdb/default.nix]: Patch fetched from GitHub applied in 49581587c20e5caa40328bf97cf22c8dbe89b2b4

**Comment by prusnak** [comment]
@GrahamcOfBorg build rocksdb

File: pkgs/development/libraries/rocksdb/default.nix (hunk 1 above)
Comment (danieldk) [reply pkgs/development/libraries/rocksdb/default.nix]: Ah, I didn't see that commit. Fetching is definitely preferred.

**Comment by danieldk** [comment]
@ofborg test ceph-single-node ceph-multi-node samba

**Review by danieldk — APPROVED** [review]
Result of `nixpkgs-review pr 93363` [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages failed to build:</summary>
<br>- bareos
<br>- sortmerna
</details>
<details>
  <summary>7 packages built:</summary>
<br>- blockbook
<br>- ceph
<br>- ceph-client
<br>- libceph
<br>- rocksdb
<br>- rocksdb_lite
<br>- samba4Full
</details>

`bareos` is already broken, `sortmerna` needs some fixes, for which I will do a PR.

I'll merge when ceph/samba tests are also successful.

**Pull request merged**